### PR TITLE
prototest: fix early return condition in AssertElementsMatch

### PR DESCRIPTION
**Pull request by rboyer**
### Description

`prototest.AssertElementsMatch` has an early return condition that returned early incorrectly when two slices were actually different.

### CODE REVIEW

File: proto/private/prototest/testing.go [hunk 1]
```diff
@@ -63,8 +70,8 @@ func AssertElementsMatch[V any](
 		}
 	}
 
-	if len(outX) == len(outY) && len(listX) == len(listY) {
-		return // matches
+	if len(outX) == len(listX) && len(outY) == len(listY) {
```
Comment (rboyer) [inline proto/private/prototest/testing.go:73]: This is the bug.

**Review by hashi-derek — APPROVED** [review]

**Comment by analogue** [comment]
Prefer the improved test in this branch https://github.com/hashicorp/consul/pull/new/spatel/better-testlist-main where `Version` doesn't have to be singled out for exclusion in the comparison.

**Review by mkeeler — APPROVED** [review]

**Review by analogue — APPROVED** [review]
🚀

**Comment by rboyer** [comment]
Since the failed tests were all in features recently changed, the backports were all done manually to elide the irrelevant changes.

**Pull request merged**